data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -96,6 +98,29 @@ public Schema requiredSchema() {
     return requiredSchema;
   }
 
+  protected int deleteMarkerIndex() {
+    if (deleteMarkerIndex != null) {
+      return deleteMarkerIndex;
+    }
+
+    int index = 0;
+    for (Types.NestedField field : requiredSchema().columns()) {
+      if (field.fieldId() != MetadataColumns.DELETE_MARK.fieldId()) {
+        index = index + 1;
```

Comment:
this feels a bit fragile to me, looks like it assumes the delete marker is the last metadata column, but if we in the future add another column it would break.